File: workload/netdp/netdp.go
// Package netdp is the workload node's eBPF datapath (the userspace half of
// workload/netdp/c/netdp.c): it loads the object, attaches the ingress program
// to the bound NIC and the egress program to each pod's host-side veth (tcx),
// and keeps the maps in step with what the agent learns — the VIP (UpdateVip),
// the l4lb fronts allowed to send IPIP (UpdateRemote), and the pod endpoints and
// their ports (the workload engine). It runs in workloadagent, non-root, with
// CAP_BPF + CAP_NET_ADMIN; it never enters a pod netns (the CNI plugin wired
// that), so it needs no CAP_SYS_ADMIN.
// See notes/ai/2026_09_25_workload_pod_netns_ebpf_design.md.
package netdp

import (
	"errors"
	"fmt"
	"log/slog"
	"net"
	"net/netip"
	"sort"
	"sync"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"
	"github.com/vishvananda/netlink"

	"github.com/on-keyday/kscale/workload"
	"github.com/on-keyday/kscale/workload/netdp/netdpmetrics"
	"github.com/on-keyday/kscale/workload/podnet"
)

// Go mirrors of the C map types. Network-order fields are byte arrays so the
// host's endianness never enters into it.
type config struct {
	VIP        [4]byte
	NICIfindex uint32
}

type portKey struct {
	Proto uint8
	Pad   uint8
	Port  [2]byte
}

type podDest struct {
	Ifindex uint32
	PodIP   [4]byte
	PodMAC  [6]byte
	HostMAC [6]byte
}

type outKey struct {
	PodIP [4]byte
	Proto uint8
	Pad   uint8
	Port  [2]byte
}

const ipprotoTCP = 6

// Counter names, in the C enum's order.
var counterNames = []string{"in_steered", "in_not_lb_src", "in_no_port", "in_err", "out_snat", "out_err"}

type Datapath struct {
	logger *slog.Logger

	mu        sync.Mutex
	coll      *ebpf.Collection
	cfg       config
	nic       string
	nicLink   link.Link
	vethLinks map[string]link.Link // host veth name -> egress tcx link
}

// Load loads the datapath object at objPath. Nothing is attached until
// BindInterface / SetEndpoints.
func Load(objPath string, logger *slog.Logger) (*Datapath, error) {
	if logger == nil {
		logger = slog.New(slog.DiscardHandler)
	}
	spec, err := ebpf.LoadCollectionSpec(objPath)
	if err != nil {
		return nil, fmt.Errorf("netdp: load %s: %w", objPath, err)
	}
	coll, err := ebpf.NewCollection(spec)
	if err != nil {
		return nil, fmt.Errorf("netdp: create collection: %w", err)
	}
	return &Datapath{logger: logger, coll: coll, vethLinks: map[string]link.Link{}}, nil
}

func (d *Datapath) Close() error {
	d.mu.Lock()
	defer d.mu.Unlock()
	var errs []error
	if d.nicLink != nil {
		errs = append(errs, d.nicLink.Close())
	}
	for _, l := range d.vethLinks {
		errs = append(errs, l.Close())
	}
	d.coll.Close()
	return errors.Join(errs...)
}

func (d *Datapath) writeConfig() error {
	return d.coll.Maps["netdp_config"].Put(uint32(0), d.cfg)
}

// BindInterface attaches the ingress program to the NIC l4lb's IPIP arrives on
// (and replies leave by). Rebinding to another NIC moves it.
func (d *Datapath) BindInterface(name string) error {
	d.mu.Lock()
	defer d.mu.Unlock()
	if name == d.nic && d.nicLink != nil {
		return nil
	}
	l, err := netlink.LinkByName(name)
	if err != nil {
		return fmt.Errorf("netdp: interface %q: %w", name, err)
	}
	tl, err := link.AttachTCX(link.TCXOptions{
		Interface: l.Attrs().Index,
		Program:   d.coll.Programs["netdp_ingress"],
		Attach:    ebpf.AttachTCXIngress,
	})
	if err != nil {
		return fmt.Errorf("netdp: attach ingress on %s: %w", name, err)
	}
	if d.nicLink != nil {
		_ = d.nicLink.Close()
	}
	d.nic, d.nicLink = name, tl
	d.cfg.NICIfindex = uint32(l.Attrs().Index)
	d.logger.Info("netdp: ingress attached", "interface", name, "ifindex", l.Attrs().Index)
	return d.writeConfig()
}

// SetVIP sets the address the datapath steers (and SNATs replies to).
func (d *Datapath) SetVIP(vip netip.Addr) error {
	if !vip.Is4() {
		return fmt.Errorf("netdp: VIP %v: only IPv4 is steered", vip)
	}
	d.mu.Lock()
	defer d.mu.Unlock()
	d.cfg.VIP = vip.As4()
	return d.writeConfig()
}

// SetLBSources replaces the set of l4lb fronts whose IPIP the ingress program
// will decap. IPIP from any other source is left to the kernel.
func (d *Datapath) SetLBSources(srcs []netip.Addr) error {
	d.mu.Lock()
	defer d.mu.Unlock()
	want := map[[4]byte]bool{}
	for _, a := range srcs {
		if a.Is4() {
			want[a.As4()] = true
		}
	}
	m := d.coll.Maps["netdp_lb_srcs"]
	return syncSet(m, want, uint8(1))
}

// SetEndpoints makes the port maps and egress attachments match eps. A port
// claimed by two endpoints goes to the first by name (eps is sorted by name) and
// the other claim is reported in the returned error; an endpoint whose host veth
// is not there yet is skipped and reported (the next sync retries).
func (d *Datapath) SetEndpoints(eps []workload.Endpoint) error {
	d.mu.Lock()
	defer d.mu.Unlock()
	var errs []error
	in := map[portKey]podDest{}
	out := map[outKey]bool{}
	veths := map[string]int{}
	owner := map[portKey]string{}
	for _, ep := range eps {
		name, err := podnet.HostVethName(ep.SandboxID)
		if err != nil {
			errs = append(errs, fmt.Errorf("%s: %w", ep.Name, err))
			continue
		}
		l, err := netlink.LinkByName(name)
		if err != nil {
			errs = append(errs, fmt.Errorf("%s: host veth %s: %w", ep.Name, name, err))
			continue
		}
		podMAC, _ := podnet.PodMAC(ep.SandboxID)
		dest := podDest{Ifindex: uint32(l.Attrs().Index), PodIP: ep.PodIP.As4()}
		copy(dest.PodMAC[:], podMAC)
		copy(dest.HostMAC[:], l.Attrs().HardwareAddr)
		veths[name] = l.Attrs().Index
		for _, p := range ep.Ports {
			k := portKey{Proto: ipprotoTCP, Port: be16(p.Port)}
			if prev, taken := owner[k]; taken {
				errs = append(errs, fmt.Errorf("%s: %s:%d already owned by %s", ep.Name, p.Proto, p.Port, prev))
				continue
			}
			owner[k] = ep.Name
			in[k] = dest
			out[outKey{PodIP: ep.PodIP.As4(), Proto: ipprotoTCP, Port: be16(p.Port)}] = true
		}
	}

	// Egress attachments first, so a pod never receives traffic it can't answer.
	for name, idx := range veths {
		if _, ok := d.vethLinks[name]; ok {
			continue
		}
		tl, err := link.AttachTCX(link.TCXOptions{
			Interface: idx,
			Program:   d.coll.Programs["netdp_egress"],
			Attach:    ebpf.AttachTCXIngress,
		})
		if err != nil {
			errs = append(errs, fmt.Errorf("attach egress on %s: %w", name, err))
			continue
		}
		d.vethLinks[name] = tl
	}
	if err := syncSet(d.coll.Maps["netdp_ports_out"], out, uint8(1)); err != nil {
		errs = append(errs, err)
	}
	if err := syncMap(d.coll.Maps["netdp_ports_in"], in); err != nil {
		errs = append(errs, err)
	}
	for name, l := range d.vethLinks {
		if _, ok := veths[name]; !ok {
			_ = l.Close() // the veth is usually gone already (CNI DEL)
			delete(d.vethLinks, name)
		}
	}
	return errors.Join(errs...)
}

// Counters sums the per-CPU counters.
func (d *Datapath) Counters() (map[string]uint64, error) {
	d.mu.Lock()
	defer d.mu.Unlock()
	out := map[string]uint64{}
	for i, name := range counterNames {
		var per []uint64
		if err := d.coll.Maps["netdp_counters"].Lookup(uint32(i), &per); err != nil {
			return nil, err
		}
		var sum uint64
		for _, v := range per {
			sum += v
		}
		out[name] = sum
	}
	return out, nil
}

// Metrics snapshots the counters and the steered-port count in the generated
// stat shape (reported via Stats and exported to prometheus). The counters live
// in unpinned maps, so they restart from zero with the agent.
func (d *Datapath) Metrics() (netdpmetrics.NetdpMetrics, error) {
	c, err := d.Counters()
	if err != nil {
		return netdpmetrics.NetdpMetrics{}, err
	}
	ports, err := d.Ports()
	if err != nil {
		return netdpmetrics.NetdpMetrics{}, err
	}
	return netdpmetrics.NetdpMetrics{
		InSteeredTotal:  c["in_steered"],
		InNotLbSrcTotal: c["in_not_lb_src"],
		InNoPortTotal:   c["in_no_port"],
		InErrTotal:      c["in_err"],
		OutSnatTotal:    c["out_snat"],
		OutErrTotal:     c["out_err"],
		SteeredPorts:    uint64(len(ports)),
	}, nil
}

// Ports lists the steered ports (for logs/tests), sorted.
func (d *Datapath) Ports() ([]string, error) {
	d.mu.Lock()
	defer d.mu.Unlock()
	var out []string
	var k portKey
	var v podDest
	it := d.coll.Maps["netdp_ports_in"].Iterate()
	for it.Next(&k, &v) {
		out = append(out, fmt.Sprintf("tcp:%d->%v", uint16(k.Port[0])<<8|uint16(k.Port[1]), net.IP(v.PodIP[:])))
	}
	sort.Strings(out)
	return out, it.Err()
}

func be16(v uint16) [2]byte { return [2]byte{byte(v >> 8), byte(v)} }

// syncSet makes a map's key set equal want, every value = val.
func syncSet[K comparable, V any](m *ebpf.Map, want map[K]bool, val V) error {
	vals := make(map[K]V, len(want))
	for k := range want {
		vals[k] = val
	}
	return syncMap(m, vals)
}

// syncMap makes a map's contents equal want: stale keys deleted, the rest put.
func syncMap[K comparable, V any](m *ebpf.Map, want map[K]V) error {
	var stale []K
	var k K
	var v V
	it := m.Iterate()
	for it.Next(&k, &v) {
		if _, ok := want[k]; !ok {
			stale = append(stale, k)
		}
	}
	if err := it.Err(); err != nil {
		return err
	}
	var errs []error
	for _, k := range stale {
		if err := m.Delete(k); err != nil && !errors.Is(err, ebpf.ErrKeyNotExist) {
			errs = append(errs, err)
		}
	}
	for k, v := range want {
		if err := m.Put(k, v); err != nil {
			errs = append(errs, err)
		}
	}
	return errors.Join(errs...)
}

File: workload/netdp/c/netdp.c
// kscale workload datapath: steer VIP:<port> TCP into pod-network containers
// and send their replies straight back (DSR). Loaded and fed by workloadagent
// (workload/netdp); see notes/ai/2026_09_25_workload_pod_netns_ebpf_design.md.
//
//   netdp_ingress  (tcx ingress on the node's bound NIC)
//     IPIP from a known l4lb front, inner TCP to VIP:<port in ports_in>
//       -> strip the outer IPv4, DNAT VIP -> pod IP, set pod/host-veth MACs,
//          bpf_redirect_peer into the pod.
//     anything else -> TC_ACT_OK (popcache's kernel IPIP decap still handles it).
//
//   netdp_egress   (tcx ingress on each pod's host-side veth = what the pod sends)
//     TCP from <pod IP, sport in ports_out>
//       -> SNAT pod IP -> VIP, bpf_redirect_neigh out of the bound NIC.
//     anything else -> TC_ACT_OK (host stack; forwarding is off, so it goes nowhere).
//
// The NAT is stateless: (proto, port) maps 1:1 to one pod, so no conntrack.
#include <linux/bpf.h>
#include <linux/if_ether.h>
#include <linux/in.h>
#include <linux/ip.h>
#include <linux/pkt_cls.h>
#include <linux/tcp.h>
#include <linux/types.h>
#include <bpf/bpf_endian.h>
#include <bpf/bpf_helpers.h>

#ifndef BPF_F_ADJ_ROOM_DECAP_L3_IPV4
#define BPF_F_ADJ_ROOM_DECAP_L3_IPV4 (1ULL << 7)
#endif

#define IP_MF 0x2000
#define IP_OFFSET 0x1FFF

struct config {
  __u32 vip;          // network order; 0 = unset (steer nothing)
  __u32 nic_ifindex;  // bound NIC for replies; 0 = unset
};

struct port_key {
  __u8 proto;
  __u8 pad;
  __u16 port;  // network order
};

struct pod_dest {
  __u32 ifindex;  // host-side veth
  __u32 pod_ip;   // network order
  __u8 pod_mac[ETH_ALEN];
  __u8 host_mac[ETH_ALEN];
};

struct out_key {
  __u32 pod_ip;  // network order
  __u8 proto;
  __u8 pad;
  __u16 port;  // network order (the pod's source port)
};

enum counter {
  C_IN_STEERED,     // decapped + redirected into a pod
  C_IN_NOT_LB_SRC,  // IPIP whose outer source is not a known l4lb (left alone)
  C_IN_NO_PORT,     // IPIP to VIP from an l4lb but no pod owns the port (left alone)
  C_IN_ERR,         // a helper failed mid-rewrite (dropped)
  C_OUT_SNAT,       // reply SNATed + redirected out
  C_OUT_ERR,
  C_MAX,
};

struct {
  __uint(type, BPF_MAP_TYPE_ARRAY);
  __uint(max_entries, 1);
  __type(key, __u32);
  __type(value, struct config);
} netdp_config SEC(".maps");

struct {
  __uint(type, BPF_MAP_TYPE_HASH);
  __uint(max_entries, 256);
  __type(key, __u32);  // l4lb front IPv4, network order
  __type(value, __u8);
} netdp_lb_srcs SEC(".maps");

struct {
  __uint(type, BPF_MAP_TYPE_HASH);
  __uint(max_entries, 1024);
  __type(key, struct port_key);
  __type(value, struct pod_dest);
} netdp_ports_in SEC(".maps");

struct {
  __uint(type, BPF_MAP_TYPE_HASH);
  __uint(max_entries, 1024);
  __type(key, struct out_key);
  __type(value, __u8);
} netdp_ports_out SEC(".maps");

struct {
  __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
  __uint(max_entries, C_MAX);
  __type(key, __u32);
  __type(value, __u64);
} netdp_counters SEC(".maps");

static __always_inline void count(__u32 c) {
  __u64 *v = bpf_map_lookup_elem(&netdp_counters, &c);
  if (v) *v += 1;
}

static __always_inline struct config *get_config(void) {
  __u32 zero = 0;
  return bpf_map_lookup_elem(&netdp_config, &zero);
}

// Offsets within an Ethernet + option-less IPv4 + TCP frame.
#define ETH_OFF_DST 0
#define ETH_OFF_SRC ETH_ALEN
#define IP_OFF (sizeof(struct ethhdr))
#define IP_OFF_CSUM (IP_OFF + __builtin_offsetof(struct iphdr, check))
#define IP_OFF_SADDR (IP_OFF + __builtin_offsetof(struct iphdr, saddr))
#define IP_OFF_DADDR (IP_OFF + __builtin_offsetof(struct iphdr, daddr))
#define TCP_OFF (IP_OFF + sizeof(struct iphdr))
#define TCP_OFF_CSUM (TCP_OFF + __builtin_offsetof(struct tcphdr, check))

static __always_inline int plain_ipv4(struct iphdr *ip) {
  return ip->version == 4 && ip->ihl == 5 && (ip->frag_off & bpf_htons(IP_MF | IP_OFFSET)) == 0;
}

SEC("tc")
int netdp_ingress(struct __sk_buff *skb) {
  void *data = (void *)(long)skb->data;
  void *data_end = (void *)(long)skb->data_end;

  struct ethhdr *eth = data;
  struct iphdr *outer = (void *)(eth + 1);
  struct iphdr *inner = (void *)(outer + 1);
  struct tcphdr *tcp = (void *)(inner + 1);
  if ((void *)(tcp + 1) > data_end) return TC_ACT_OK;
  if (eth->h_proto != bpf_htons(ETH_P_IP)) return TC_ACT_OK;
  if (!plain_ipv4(outer) || outer->protocol != IPPROTO_IPIP) return TC_ACT_OK;

  if (!bpf_map_lookup_elem(&netdp_lb_srcs, &outer->saddr)) {
    count(C_IN_NOT_LB_SRC);
    return TC_ACT_OK;
  }
  struct config *cfg = get_config();
  if (!cfg || cfg->vip == 0) return TC_ACT_OK;
  if (!plain_ipv4(inner) || inner->protocol != IPPROTO_TCP || inner->daddr != cfg->vip) return TC_ACT_OK;

  struct port_key key = {.proto = IPPROTO_TCP, .port = tcp->dest};
  struct pod_dest *dst = bpf_map_lookup_elem(&netdp_ports_in, &key);
  if (!dst) {
    count(C_IN_NO_PORT);
    return TC_ACT_OK;
  }
  __u32 vip = cfg->vip;
  __u32 pod_ip = dst->pod_ip;
  __u32 ifindex = dst->ifindex;
  __u8 macs[2 * ETH_ALEN];
  __builtin_memcpy(macs, dst->pod_mac, ETH_ALEN);
  __builtin_memcpy(macs + ETH_ALEN, dst->host_mac, ETH_ALEN);

  // Strip the outer IPv4 header (the Ethernet header stays in front).
  if (bpf_skb_adjust_room(skb, -(int)sizeof(struct iphdr), BPF_ADJ_ROOM_MAC, BPF_F_ADJ_ROOM_DECAP_L3_IPV4)) goto err;
  // DNAT VIP -> pod IP, fixing the IPv4 and the TCP (pseudo-header) checksums.
  if (bpf_l3_csum_replace(skb, IP_OFF_CSUM, vip, pod_ip, sizeof(pod_ip))) goto err;
  if (bpf_l4_csum_replace(skb, TCP_OFF_CSUM, vip, pod_ip, BPF_F_PSEUDO_HDR | sizeof(pod_ip))) goto err;
  if (bpf_skb_store_bytes(skb, IP_OFF_DADDR, &pod_ip, sizeof(pod_ip), 0)) goto err;
  if (bpf_skb_store_bytes(skb, ETH_OFF_DST, macs, sizeof(macs), 0)) goto err;

  count(C_IN_STEERED);
  return bpf_redirect_peer(ifindex, 0);
err:
  count(C_IN_ERR);
  return TC_ACT_SHOT;
}

SEC("tc")
int netdp_egress(struct __sk_buff *skb) {
  void *data = (void *)(long)skb->data;
  void *data_end = (void *)(long)skb->data_end;

  struct ethhdr *eth = data;
  struct iphdr *ip = (void *)(eth + 1);
  struct tcphdr *tcp = (void *)(ip + 1);
  if ((void *)(tcp + 1) > data_end) return TC_ACT_OK;
  if (eth->h_proto != bpf_htons(ETH_P_IP)) return TC_ACT_OK;
  if (!plain_ipv4(ip) || ip->protocol != IPPROTO_TCP) return TC_ACT_OK;

  struct out_key key = {.pod_ip = ip->saddr, .proto = IPPROTO_TCP, .port = tcp->source};
  if (!bpf_map_lookup_elem(&netdp_ports_out, &key)) return TC_ACT_OK;
  struct config *cfg = get_config();
  if (!cfg || cfg->vip == 0 || cfg->nic_ifindex == 0) return TC_ACT_OK;
  __u32 vip = cfg->vip;
  __u32 pod_ip = ip->saddr;
  __u32 nic = cfg->nic_ifindex;

  // SNAT pod IP -> VIP so the client sees the reply come from the address it
  // dialled (DSR: the reply does not go back through l4lb).
  if (bpf_l3_csum_replace(skb, IP_OFF_CSUM, pod_ip, vip, sizeof(vip))) goto err;
  if (bpf_l4_csum_replace(skb, TCP_OFF_CSUM, pod_ip, vip, BPF_F_PSEUDO_HDR | sizeof(vip))) goto err;
  if (bpf_skb_store_bytes(skb, IP_OFF_SADDR, &vip, sizeof(vip), 0)) goto err;

  count(C_OUT_SNAT);
  // FIB lookup + neighbor resolution fill in the L2 header for the NIC.
  return bpf_redirect_neigh(nic, NULL, 0, 0);
err:
  count(C_OUT_ERR);
  return TC_ACT_SHOT;
}

char _license[] SEC("license") = "GPL";
